posts/navarroaxel/13.running-graphics-apps-in-docker.md

```diff
@@ -0,0 +1,138 @@
+```
+
+We need to share the `.X11-unix/` directory where the connection [socket](https://en.wikipedia.org/wiki/Unix_domain_socket) is located, and `DISPLAY` is the name of the display of the host.
+
+The `--network=host` argument makes the container use the same network of the host, and the container does not get its own IP address. This avoids the container being network isolated. 🪛
+
+## Starting the GUI application
+
+Before starting the app we must authenticate the X Window System of the container with the host.
+
+Get the authorization entry of the host using the `xauth` command:
+
+```bash
+xauth list
+```
+
+Copy the output of the command and run this in the container (replacing the placeholders with the values from the clipboard):
+
+```bash
+xauth add <display_name> <protocol_name> <hex_key>
+```
+
+Now you are able to start GUI applications inside the container! 🎉
+
+```bash
+/opt/workspacesclient/workspacesclient
+```
+
+## Alternative: use VNC instead of X11
+
+We need to make a few changes to run a VNC server in the container:
+
+```Dockerfile
+FROM ubuntu:focal
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt update && \
+  apt install -y libusb-1.0-0 gpg wget x11vnc xvfb && \
+  wget -q -O - https://workspaces-client-linux-public-key.s3-us-west-2.amazonaws.com/ADB332E7.asc | apt-key add - && \
+  echo "deb [arch=amd64] https://d3nt0h4h6pmmc4.cloudfront.net/ubuntu focal main" > /etc/apt/sources.list.d/amazon-workspaces-clients.list && \
+  apt remove -y wget && \
+  apt update && \
+  apt install -y workspacesclient && \
+  rm -rf /var/lib/apt/lists/*
+
+RUN echo "exec /opt/workspacesclient/workspacesclient" > ~/.xinitrc && \
+  chmod +x ~/.xinitrc
+
+CMD ["x11vnc", "-create", "-forever"]
+```
+
+The `x11vnc` command will start an X11 session in the container launching the application specified in the `~/.xinitrc` file.
+
+```bash
+docker build -t workspace .
+```
+
+Then, we can run the container
+
+```bash
+docker run --name ws -it workspace
+```
+
+But how can we know the IP address of the container? Just check it with the following command:
+
+```bash
+docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' ws
+```
+
+Now you can connect to the VNC server. 🔌
+
+## Conclusion
+
+We can run graphic applications that work in specific Linux distros using Docker 🐋, by connecting directly to our X Window System. This uses less memory and disk space than a VM.
+
+You can get the same in Windows as the host using [Xming](https://sourceforge.net/projects/xming/).
+
+`xauth` could be used in conjunction with SSH to run remote GUI apps 😉 in our local, but that's for another post.
+
+Alternatively, you can start a VNC server in a container allowing more flexibility either if you use macOS, or if you can't connect using X11 because you're using [Wayland](https://en.wikipedia.org/wiki/Wayland_(display_server_protocol)). But this doesn't file like a native window on your OS.
```

Comment:
```suggestion
Alternatively, you can start a VNC server in a container allowing more flexibility either if you use macOS, or if you can't connect using X11 because you're using [Wayland](https://en.wikipedia.org/wiki/Wayland_(display_server_protocol)). But this doesn't feel like a native window on your OS.
```